Type-hinted additional-builtins raise error since 2.12

### Bug description

```python
# pylint: disable=missing-docstring

from typing import TYPE_CHECKING, Any, Dict

if TYPE_CHECKING:
    __additional_builtin__: Dict[str, Any]


def run():
    return __additional_builtin__["test"]
```

This example will warn about `__additional_builtin__` being an `undefined-variable` since 2.12.0 despite `__additional_builtin__` being listed in `.pylintrc`.

Without the type hints, no warning is issued.

I looked at the source code, and tested it locally, and the new warning is added here: https://github.com/PyCQA/pylint/blob/main/pylint/checkers/variables.py#L1683-L1685=

The `__additional_builtin__` variables is correctly identified as only previously assigned as a type hint, but the branch does not check for `VARIABLES.additional-builtins` as, for example, here: https://github.com/PyCQA/pylint/blob/main/pylint/checkers/variables.py#L1438=

---

I am developing a few plugins, e.g. for saltstack/salt, which are invoked with additional built ins by salt. Pylint is used in CI and the version maintained by renovate-bot. That highlighted the first failures here: https://github.com/jgraichen/salt-tower/pull/25

### Configuration

```ini
[VARIABLES]
additional-builtins=__additional_builtin__
```


### Command used

```shell
pylint test.py
```


### Pylint output

```shell
************* Module test
test.py:10:11: E0602: Undefined variable '__additional_builtin__' (undefined-variable)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

```

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### Pylint version

```shell
pylint 2.13.5
astroid 2.11.2
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
```
### OS / Environment

Debian 11

### Additional dependencies

_No response_

--- a/ChangeLog
+++ b/ChangeLog
@@ -211,9 +211,6 @@ Release date: TBA
 
   Closes #6388
 
-* Fixed a false positive for ``redefined-builtin`` for builtins specified in
-  ``--allowed-redefined-builtins`` and redefined at a module scope.
-
 
 What's New in Pylint 2.13.5?
 ============================

--- a/ChangeLog
+++ b/ChangeLog
@@ -211,6 +211,9 @@ Release date: TBA
 
   Closes #6388
 
+* Fixed a false positive for ``redefined-builtin`` for builtins specified in
+  ``--allowed-redefined-builtins`` and redefined at a module scope.
+
 
 What's New in Pylint 2.13.5?
 ============================

--- a/ChangeLog
+++ b/ChangeLog
@@ -206,6 +206,11 @@ Release date: TBA
   Closes #5815
   Closes #5406
 
+* Fixed a false positive for ``unused-variable`` when a builtin specified in
+  ``--additional-builtins`` is given a type annotation.
+
+  Closes #6388
+
 
 What's New in Pylint 2.13.5?
 ============================

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -592,9 +592,6 @@ Other Changes
 
   Closes #6388
 
-* Fixed a false positive for ``redefined-builtin`` for builtins specified in
-  ``--allowed-redefined-builtins`` and redefined at a module scope.
-
 * Fix false positive for 'nonexistent-operator' when repeated '-' are
   separated (e.g. by parens).
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -592,6 +592,9 @@ Other Changes
 
   Closes #6388
 
+* Fixed a false positive for ``redefined-builtin`` for builtins specified in
+  ``--allowed-redefined-builtins`` and redefined at a module scope.
+
 * Fix false positive for 'nonexistent-operator' when repeated '-' are
   separated (e.g. by parens).
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -587,6 +587,11 @@ Other Changes
 
   Closes #5394
 
+* Fixed a false positive for ``unused-variable`` when a builtin specified in
+  ``--additional-builtins`` is given a type annotation.
+
+  Closes #6388
+
 * Fix false positive for 'nonexistent-operator' when repeated '-' are
   separated (e.g. by parens).
 

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1121,8 +1121,6 @@ class VariablesChecker(BaseChecker):
 
         for name, stmts in node.locals.items():
             if utils.is_builtin(name):
-                if self._allowed_redefined_builtin(name):
-                    continue
                 if self._should_ignore_redefined_builtin(stmts[0]) or name == "__doc__":
                     continue
                 self.add_message("redefined-builtin", args=name, node=stmts[0])

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1121,6 +1121,8 @@ class VariablesChecker(BaseChecker):
 
         for name, stmts in node.locals.items():
             if utils.is_builtin(name):
+                if self._allowed_redefined_builtin(name):
+                    continue
                 if self._should_ignore_redefined_builtin(stmts[0]) or name == "__doc__":
                     continue
                 self.add_message("redefined-builtin", args=name, node=stmts[0])
@@ -2066,7 +2068,9 @@ class VariablesChecker(BaseChecker):
         if not isinstance(defstmt, nodes.AnnAssign) or defstmt.value:
             return False
 
-        if node.name in self.linter.config.additional_builtins:
+        if node.name in self.linter.config.additional_builtins or utils.is_builtin(
+            node.name
+        ):
             return False
 
         defstmt_frame = defstmt.frame(future=True)

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2059,10 +2059,14 @@ class VariablesChecker(BaseChecker):
             or any(isinstance(arg, nodes.IfExp) for arg in value.args)
         )
 
-    @staticmethod
-    def _is_only_type_assignment(node: nodes.Name, defstmt: nodes.Statement) -> bool:
+    def _is_only_type_assignment(
+        self, node: nodes.Name, defstmt: nodes.Statement
+    ) -> bool:
         """Check if variable only gets assigned a type and never a value."""
         if not isinstance(defstmt, nodes.AnnAssign) or defstmt.value:
+            return False
+
+        if node.name in self.linter.config.additional_builtins:
             return False
 
         defstmt_frame = defstmt.frame(future=True)

--- a/tests/functional/u/undefined/undefined_variable_typing.py
+++ b/tests/functional/u/undefined/undefined_variable_typing.py
@@ -16,6 +16,8 @@ c = ...  # type: Bar.Boo
 
 if TYPE_CHECKING:
     __additional_builtin__: Dict[str, Any]
+    # For why this would emit redefined-builtin: https://github.com/PyCQA/pylint/pull/3643
+    # pylint: disable-next=redefined-builtin
     repr: Any
 
 

--- a/tests/functional/u/undefined/undefined_variable_typing.py
+++ b/tests/functional/u/undefined/undefined_variable_typing.py
@@ -16,8 +16,10 @@ c = ...  # type: Bar.Boo
 
 if TYPE_CHECKING:
     __additional_builtin__: Dict[str, Any]
+    repr: Any
 
 
 def run():
     """https://github.com/PyCQA/pylint/issues/6388"""
+    print(repr)
     return __additional_builtin__["test"]

--- a/tests/functional/u/undefined/undefined_variable_typing.py
+++ b/tests/functional/u/undefined/undefined_variable_typing.py
@@ -4,9 +4,20 @@
 # Ensure attribute lookups in type comments are accounted for.
 # Reported in https://github.com/PyCQA/pylint/issues/4603
 
+from typing import TYPE_CHECKING, Any, Dict
+
 import foo
 from foo import Bar, Boo
 
 a = ...  # type: foo.Bar
 b = ...  # type: foo.Bar[Boo]
 c = ...  # type: Bar.Boo
+
+
+if TYPE_CHECKING:
+    __additional_builtin__: Dict[str, Any]
+
+
+def run():
+    """https://github.com/PyCQA/pylint/issues/6388"""
+    return __additional_builtin__["test"]

--- a/tests/functional/u/undefined/undefined_variable_typing.rc
+++ b/tests/functional/u/undefined/undefined_variable_typing.rc
@@ -2,4 +2,3 @@
 except_implementations=PyPy
 [VARIABLES]
 additional-builtins=__additional_builtin__
-allowed-redefined-builtins=repr

--- a/tests/functional/u/undefined/undefined_variable_typing.rc
+++ b/tests/functional/u/undefined/undefined_variable_typing.rc
@@ -2,3 +2,4 @@
 except_implementations=PyPy
 [VARIABLES]
 additional-builtins=__additional_builtin__
+allowed-redefined-builtins=repr

--- a/tests/functional/u/undefined/undefined_variable_typing.rc
+++ b/tests/functional/u/undefined/undefined_variable_typing.rc
@@ -1,2 +1,4 @@
 [testoptions]
 except_implementations=PyPy
+[VARIABLES]
+additional_builtins=__additional_builtin__

--- a/tests/functional/u/undefined/undefined_variable_typing.rc
+++ b/tests/functional/u/undefined/undefined_variable_typing.rc
@@ -1,4 +1,4 @@
 [testoptions]
 except_implementations=PyPy
 [VARIABLES]
-additional_builtins=__additional_builtin__
+additional-builtins=__additional_builtin__

